### PR TITLE
pkg/daemon: drop nodeWriter interface + set resourceVersion when patching

### PR DESCRIPTION
**Pull request by runcom**
tl;dr; refer to https://github.com/openshift/machine-config-operator/pull/1923#issuecomment-657735033 for a detailed diagnosis of the issue

These patches do two (major) things:

- remove the nodeWriter interface, not needed now that the daemon runs as a real controller and we can afford to fail and retry
- set node's resourceVersion when patching, making sure we get a 409 on conflict and retry

Minor changes in different commits:

- react to Reason annotation changes (why we never did that is a unknown to me, so far, we're sticking to an old degraded reason if the reason changes but we're still degraded, it's just wrong...)
- use a faster workqueue for the render controller (see commit message specifically)

Long story:

The second point above is crucial as it's the source of a bug, let me try to explain what happens:

- resync on an informer means: replay the in-memory state that I have.  It does not mean relist from the apiserver
- the lister.get is based on the cached state, so if the watch hasn't observed the change, the lister.Get won't have it

Given the assumptions above (thanks deads for explaining):

- the daemon update the node object when it degrades
- it immediately retries the action, lister.get yielding an _old_ cached state
- that defeats guards like "switch to working only when not degraded"

Explanation to last above above: because of how we patch node objects, we may end up checking
the state annotation with an _old_ object (passing the guard) but then updating the right object
with states that don't make sense anymore (like a Working state and a real Degraded reason).

The fix for all that madness is pretty simple (after speaking to deads):

- use resourceVersion when patching the node object

That makes sure that if we try to patch an object using an old resouce version we'll get a conflict error,
we then exit the guard we entered, and retry with a fresher object, all good.

This can be tested with (credits to deads again)

- oc get ns/foo -ojson
- build your patch with RV
- oc annotate ns/foo something-to-change-rv=true
- oc patch with RV

step four should get a conflict error

Signed-off-by: Antonio Murdaca <runcom@linux.com>

### CODE REVIEW

**Comment by runcom** [comment]
And here's the race being taken care of (and we'll just retry):
```
I0714 09:45:48.563897    2028 kubeutils.go:58] runcom patch bytes: {"metadata":{"annotations":{"machineconfiguration.openshift.io/reason":"failed to run pivot: failed to run pivot: exit status 1: failed to run command podman (6 tries): timed out waiting for the condition: running podman pull -q --authfile /var/lib/kubelet/config.json registry.svc.ci.openshift.org/ocp/4.6@sha256:d746ebe14faaa6d1a1847c0d939373b74306c38c39f77315e559a7ec676fbbb8 failed: Error: error pulling image \"registry.svc.ci.openshift.org/ocp/4.6@sha256:d746ebe14faaa6d1a1847c0d939373b74306c38c39f77315e559a7ec676fbbb8\": unable to pull registry.svc.ci.openshift.org/ocp/4.6@sha256:d746ebe14faaa6d1a1847c0d939373b74306c38c39f77315e559a7ec676fbbb8: unable to pull image: Error initializing source docker://registry.svc.ci.openshift.org/ocp/4.6@sha256:d746ebe14faaa6d1a1847c0d939373b74306c38c39f77315e559a7ec676fbbb8: Error reading manifest sha256:d746ebe14faaa6d1a1847c0d939373b74306c38c39f77315e559a7ec676fbbb8 in registry.svc.ci.openshift.org/ocp/4.6: manifest unknown: manifest unknown\n: exit status 125","machineconfiguration.openshift.io/state":"Degraded"},"resourceVersion":"40664"}}
I0714 09:45:48.577978    2028 kubeutils.go:58] runcom patch bytes: {"metadata":{"resourceVersion":"40664"}}
E0714 09:45:48.587757    2028 writer.go:55] Marking Degraded due to: error setting node's state to Working: Operation cannot be fulfilled on nodes "ci-ln-s0rhzmb-f76d1-nlfzr-worker-b-57j9w": the object has been modified; please apply your changes to the latest version and try again
I0714 09:45:48.589432    2028 kubeutils.go:58] runcom patch bytes: {"metadata":{"annotations":{"machineconfiguration.openshift.io/reason":"error setting node's state to Working: Operation cannot be fulfilled on nodes \"ci-ln-s0rhzmb-f76d1-nlfzr-worker-b-57j9w\": the object has been modified; please apply your changes to the latest version and try again"},"resourceVersion":"43416"}}
```
/hold cancel

File: internal/kubeutils.go [hunk 1]
```diff
@@ -22,33 +22,37 @@ import (
...
-		}
-
-		node, err = client.Patch(context.TODO(), nodeName, types.StrategicMergePatchType, patchBytes, metav1.PatchOptions{})
-		return err
+		var innerErr error
+		node, innerErr = UpdateNode(client, lister, nodeName, f)
+		return innerErr
 	}); err != nil {
 		// may be conflict if max retries were hit
 		return nil, fmt.Errorf("unable to update node %q: %v", node, err)
```
Comment (ericavonb) [inline internal/kubeutils.go:30]: Can we use `nodeName` rather than the node object in the error message here? Since `node` could be nil if we hit max retries (also putting the whole node object into the error has caused issues before)

File: internal/kubeutils.go [hunk 2]
```diff
@@ -22,33 +22,37 @@ import (
...
+	if err != nil {
+		return nil, err
+	}
+	nodeClone := n.DeepCopy()
+	f(nodeClone)
+	newNode, err := json.Marshal(nodeClone)
+	if err != nil {
+		return nil, err
+	}
+	patchBytes, err := strategicpatch.CreateTwoWayMergePatch(oldNode, newNode, corev1.Node{})
```
Comment (ericavonb) [inline internal/kubeutils.go:53]: Wait, why do we marshal to json if we immediately unmarshal it? Something to do with schemas and versioning?

File: pkg/controller/render/render_controller.go [hunk 3]
```diff
@@ -84,8 +85,10 @@ func New(
 	ctrl := &Controller{
 		client:        mcfgClient,
 		eventRecorder: eventBroadcaster.NewRecorder(scheme.Scheme, corev1.EventSource{Component: "machineconfigcontroller-rendercontroller"}),
-		queue:         workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "machineconfigcontroller-rendercontroller"),
 	}
+	ctrl.queue = workqueue.NewNamedRateLimitingQueue(workqueue.NewMaxOfRateLimiter(
+		&workqueue.BucketRateLimiter{Limiter: rate.NewLimiter(rate.Limit(5*time.Second), 1)},
```
Comment (ericavonb) [inline pkg/controller/render/render_controller.go:90]: Can we make these rate limits a const with a descriptive comment about why e.g. 5 sec was chosen? like we do for `renderDelay` above

File: internal/kubeutils.go (hunk 1 above)
Comment (runcom) [reply internal/kubeutils.go:30]: great catch! thanks Erica!

File: internal/kubeutils.go (hunk 2 above)
Comment (runcom) [reply internal/kubeutils.go:53]: pretty sure that's the case when dealing with `CreateTwoWayMergePatch` yeah 🤷‍♂️

File: pkg/daemon/daemon.go [hunk 4]
```diff
@@ -372,11 +367,18 @@ func (dn *Daemon) handleErr(err error, key interface{}) {
...
 func (dn *Daemon) updateErrorState(err error) {
-	switch errors.Cause(err) {
+	cause := errors.Cause(err)
+	switch cause {
 	case errUnreconcilable:
-		dn.nodeWriter.SetUnreconcilable(err, dn.kubeClient.CoreV1().Nodes(), dn.nodeLister, dn.name)
+		dn.setUnreconcilable(err)
 	default:
-		dn.nodeWriter.SetDegraded(err, dn.kubeClient.CoreV1().Nodes(), dn.nodeLister, dn.name)
+		if apierrors.IsConflict(cause) {
```
Comment (ericavonb) [inline pkg/daemon/daemon.go:375]: nit: rather than nest an if clause in the switch, remove the switch.
e.g.
```go
	cause := errors.Cause(err)
	if cause == errUnreconcilable {
		return dn.setUnreconcilable(err)
	}
	if apierrors.IsConflict(cause) {
		// skip conflict errors, those are retried and should never result in a permanent
		// failure worth reporting to the node annoation reasons
		glog.Infof("API conflict error: %v, retrying", err)
		return
	}
	dn.setDegraded(err)
}
```

**Comment by runcom** [comment]
@ericavonb addressed your comments, can you ptal 👼

**Comment by runcom** [comment]
I really want a test for this to make sure the error wrapping + conditionals to avoid reporting degraded always work (bear with me)

/hold

File: pkg/daemon/node.go [hunk 5]
```diff
@@ -43,7 +43,7 @@ func (dn *Daemon) loadNodeAnnotations(node *corev1.Node) (*corev1.Node, error) {
 	}
 
 	glog.Infof("Setting initial node config: %s", initial[constants.CurrentMachineConfigAnnotationKey])
-	n, err := setNodeAnnotations(dn.kubeClient.CoreV1().Nodes(), dn.nodeLister, node.Name, initial)
+	n, err := dn.setNodeAnnotations(initial)
```
Comment (yuqi-zhang) [inline pkg/daemon/node.go:46]: hm, do you know why for some instances we don't use the retry version of setNodeAnnotations (e.g. here)?

File: pkg/daemon/daemon.go [hunk 6]
```diff
@@ -372,12 +367,18 @@ func (dn *Daemon) handleErr(err error, key interface{}) {
...
-	default:
-		dn.nodeWriter.SetDegraded(err, dn.kubeClient.CoreV1().Nodes(), dn.nodeLister, dn.name)
+	cause := errors.Cause(err)
+	if cause == errUnreconcilable {
+		dn.setUnreconcilable(err)
+		return
+	}
+	if apierrors.IsConflict(cause) {
+		// skip conflict errors, those are retried and should never result in a permanent
+		// failure worth reporting to the node annoation reasons
```
Comment (yuqi-zhang) [inline pkg/daemon/daemon.go:377]: nit `annoation` -> `annotation`

File: internal/kubeutils.go [hunk 7]
```diff
@@ -22,33 +22,37 @@ import (
...
 }
+
+// UpdateNode calls f to update a node object in Kubernetes.
+func UpdateNode(client corev1client.NodeInterface, lister corev1lister.NodeLister, nodeName string, f func(*corev1.Node)) (*corev1.Node, error) {
+	n, err := lister.Get(nodeName)
+	if err != nil {
+		return nil, err
+	}
+	oldNodeClone := n.DeepCopy()
+	oldNodeClone.ResourceVersion = ""
```
Comment (yuqi-zhang) [inline internal/kubeutils.go:42]: So in the PR you mentioned that:
```
That makes sure that if we try to patch an object using an old resouce version we'll get a conflict error,
we then exit the guard we entered, and retry with a fresher object, all good.
```

I'm a little confused as to how we're actually performing this. By setting the old `ResourceVersion` to empty, the merge should set the ResourceVersion to whatever is in newNode, right? Is that different from before? I don't really see a difference in how ResourceVersion is used I guess is my question.

Comment (runcom) [reply internal/kubeutils.go:42]: Right, let me explain better the situation here and maybe you can help me word this better Jerry:

Before (just an example, annotations is a map, but I'm doing array for simplicity):

- grab the node object (we'll call it `instance{number: 1, resourceVersion: 42, annotations: ["test1"]}`)
- clone it (now we have `instance{number: 2, resourceVersion: 42, annotations: ["test1"]}`)
- modify the annotations on the clone (now we have `instance{instance{number: 2, resourceVersion: 42, annotations: ["test1", "test2"]}`)
- run the instances through `CreateTwoWayMergePatch`

What `CreateTwoWayMergePatch` does and how Patch works is crucial here. When `CreateTwoWayMergePatch` is run above the following (truncated) `diff` is generated:

```
node
  annotations: ["test1", "test2"]
```

You can notice the above generated a diff, meaning it's giving you just the diff object that we'll use to patch. The result is that Patching with the diff above will only ever change annotations, as that's effectively the only changes `CreateTwoWayMergePatch` can see from the two objects.

The fix I did here forces us to always generate **also** a resourceVersion diff so that it can be noticed at Patch time and the apiserver complains if it's old.

One other thing to note here is that we're not playing with the original `n` node instance, only `oldNodeClone` and `oldNode` to generate the diff (`n` remains untouched)

By setting resourceVersion to empty, we'll go like this:

- grab the node object (we'll call it `instance{number: 1, resourceVersion: 42, annotations: ["test1"]}`)
- make a first clone and empty out the resourceVersion `instance{number: 1, resourceVersion: "", annotations: ["test1"]}`
- clone the original object again (now we have `instance{number: 2, resourceVersion: 42, annotations: ["test1"]}`)
- modify the annotations on this last clone (now we have `instance{instance{number: 2, resourceVersion: 42, annotations: ["test1", "test2"]}`)
- run the instances through `CreateTwoWayMergePatch`

Now, the diff we get is:

```
node
  annotations: ["test1", "test2"]
  resourceVersion: 42
``` 

And we'll patch with the above. If we send `42` but the apiserver already knows about `43` we'll get a conflict

File: pkg/daemon/node.go (hunk 5 above)
Comment (runcom) [reply pkg/daemon/node.go:46]: Yep, for the critical path (from `syncNode` to reboot) we don't really want to retry, it's the controller mechanism which will retry for us. For things like setting the SSH annotation we don't really care, I could have _not_ add a retry as well as it's retried anyway, but it looked more consistent there to retry 🤷‍♂️

File: internal/kubeutils.go (hunk 7 above)
Comment (runcom) [reply internal/kubeutils.go:42]: Right, let me explain better the situation here and maybe you can help me word this better Jerry:
Before (just an example, annotations is a map, but I'm doing array for simplicity):
- grab the node object (we'll call it `instance{number: 1, resourceVersion: 42, annotations: ["test1"]}`)
- clone it (now we have `instance{number: 2, resourceVersion: 42, annotations: ["test1"]}`)
- modify the annotations on the clone (now we have `instance{instance{number: 2, resourceVersion: 42, annotations: ["test1", "test2"]}`)
- run the instances through `CreateTwoWayMergePatch`
What `CreateTwoWayMergePatch` does and how Patch works is crucial here. When `CreateTwoWayMergePatch` is run above the following (truncated) `diff` is generated:
```
node
  annotations: ["test1", "test2"]
```
You can notice the above generated a diff, meaning it's giving you just the diff object that we'll use to patch. The result is that Patching with the diff above will only ever change annotations, as that's effectively the only changes `CreateTwoWayMergePatch` can see from the two objects.

The fix I did here forces us to always generate **also** a resourceVersion diff so that it can be noticed at Patch time and the apiserver complains if it's old.
One other thing to note here is that we're not playing with the original `n` node instance, only `oldNodeClone` and `oldNode` to generate the diff (`n` remains untouched)
By setting resourceVersion to empty, we'll go like this:
- grab the node object (we'll call it `instance{number: 1, resourceVersion: 42, annotations: ["test1"]}`)
- make a first clone and empty out the resourceVersion `instance{number: 1, resourceVersion: "", annotations: ["test1"]}`
- clone the original object again (now we have `instance{number: 2, resourceVersion: 42, annotations: ["test1"]}`)
- modify the annotations on this last clone (now we have `instance{instance{number: 2, resourceVersion: 42, annotations: ["test1", "test2"]}`)
- run the first old clone and the other clone instances through `CreateTwoWayMergePatch`
Now, the diff we get is:
```
node
  annotations: ["test1", "test2"]
  resourceVersion: 42
``` 

It is ultimately making sure that if we're trying to patch an object, at least the changes we're applying are applied to the very last version of an object in the apiserver, otherwise, we'll retry.

File: pkg/daemon/node.go (hunk 5 above)
Comment (runcom) [reply pkg/daemon/node.go:46]: Yep, for the critical path (from `syncNode` to reboot) we don't really want to retry, it's the controller mechanism which will retry for us. For things like setting the SSH annotation we don't really care, I could have _not_ add a retry as well as it's retried anyway, but it looked more consistent there to retry 🤷‍♂️

File: internal/kubeutils.go (hunk 7 above)
Comment (yuqi-zhang) [reply internal/kubeutils.go:42]: Ah, I see, no that makes perfect sense. Awesome explanation!

Maybe for future purposes it'd be nice to add a comment like:

`// Add empty resourceVersion, so CreateTwoWayMergePatch returns a diff with resourceVersion of the node object when comparing.`

Comment (runcom) [reply internal/kubeutils.go:42]: Adding that, that's really helpful Jerry, thanks for pointing it out!

**Comment by cgwalters** [comment]
Wouldn't it be sufficient to solve the bug to do *either* of these things (drop nodeWriter or retry on resourceVersion) - basically the core problem is we're racing against our own process, so if we patched the node synchronously there wouldn't be a race anymore.  Right?
